### PR TITLE
[TIMOB-25886] Refactor Ti.App._restart()

### PR DESCRIPTION
**Pull request by garymathews**
- Improve Ti.App._restart() performance
##### NOTES
- ~Need to test with native Titanium modules~ Works
- No tests as it restarts the context

##### TEST CASE
```JS
var win = Ti.UI.createWindow({ backgroundColor: 'gray', layout: 'vertical' }),
    globalBtn = Ti.UI.createButton({ title: 'CHECK GLOBAL (always undefined)' }),
    restartBtn = Ti.UI.createButton({ title: 'RESTART' });

globalBtn.addEventListener('click', function () {
    alert(JSON.stringify(global.TEST));
});

restartBtn.addEventListener('click', function () {
    global.TEST = true;
    Ti.App._restart();
});

setTimeout(() => {
    win.backgroundColor = 'green';
}, 3000);

win.add([ globalBtn, restartBtn ]);
win.open();
```

[JIRA Ticket](https://jira.appcelerator.org/browse/TIMOB-25886)

### CODE REVIEW

File: android/modules/app/src/java/ti/modules/titanium/app/AppModule.java [hunk 1]
```diff
@@ -213,14 +217,23 @@ public boolean getAccessibilityEnabled()
...
+
+		// allow termination again
+		TiBaseActivity.canFinishRoot = true;
+
+		// restart kroll runtime
+		runtime.doDispose();
+		runtime.initRuntime();
+
+		// manually re-launch app
+		runtime.doRunModule(KrollAssetHelper.readAsset(APP_PATH), APP_PATH, getActivityProxy());
```
Comment (jquick-axway) [inline android/modules/app/src/java/ti/modules/titanium/app/AppModule.java]: Our `TiLaunchActivity.loadActivityScript()` and the `resolveUrl()` method it calls does some extra logic when loading a JavaScript on startup. Particularly with Alloy. We might need to do a code shift.
https://github.com/appcelerator/titanium_mobile/blob/master/android/titanium/src/java/org/appcelerator/titanium/TiLaunchActivity.java#L120

File: android/modules/app/src/java/ti/modules/titanium/app/AppModule.java [hunk 2]
```diff
@@ -213,14 +217,23 @@ public boolean getAccessibilityEnabled()
...
-		Application app = (Application) KrollRuntime.getInstance().getKrollApplication();
-		Intent i = app.getPackageManager().getLaunchIntentForPackage(app.getPackageName());
-		i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-		i.addFlags(Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
-		i.addCategory(Intent.CATEGORY_LAUNCHER);
-		i.setAction(Intent.ACTION_MAIN);
+		KrollRuntime runtime = KrollRuntime.getInstance();
+
+		// prevent termination of root activity via TiBaseActivity.shouldFinishRootActivity()
+		TiBaseActivity.canFinishRoot = false;
```
Comment (jquick-axway) [inline android/modules/app/src/java/ti/modules/titanium/app/AppModule.java]: Once we set `canFinishRoot` to `false`, then backing out of the first Ti.UI.Window in a restarted window will now show the launcher activity, right? We may need to reset `canFinishRoot` back to `true` after the restart.

Perhaps this shouldn't be an exposed public global and handle it via a method.

Comment (garymathews) [reply android/modules/app/src/java/ti/modules/titanium/app/AppModule.java]: `canFinishRoot` returns to true on [Line 229](https://github.com/appcelerator/titanium_mobile/pull/9946/files#diff-6605aa182d97e7cfaab1dd4f36bcf791R229)

File: android/modules/app/src/java/ti/modules/titanium/app/AppModule.java (hunk 1 above)
Comment (garymathews) [reply android/modules/app/src/java/ti/modules/titanium/app/AppModule.java]: Ah yes, I'll test these changes with an Alloy app

Comment (garymathews) [reply android/modules/app/src/java/ti/modules/titanium/app/AppModule.java]: Update, works fine with Alloy apps

**Comment by sgtcoolguy** [comment]
@jquick-axway ping

**Pull request merged**